File: Medium/377. Combination Sum IV/Solution.py
class Solution(object):
    def __init__ (self):
        self.dp = []

    def solve(self, nums, target, val):
        if val > target:
            return 0
        if val == target:
            return 1

        if self.dp[val] != -1:
            return self.dp[val]

        ret = 0
        
        for it in nums:
            ret += self.solve(nums, target, val + it)

        self.dp[val] = ret
        return ret


    def combinationSum4(self, nums, target):
        self.dp = [-1] * (target)
        return self.solve(nums, target, 0)

File: Medium/377. Combination Sum IV/Solution.java
import java.util.Arrays;

class Solution {
    int[] dp;

    int solve(int[] nums, int target, int val) {
        if (val > target)
            return 0;
        if (val == target)
            return 1;

        if (dp[val] != -1)
            return dp[val];

        int ret = 0;
        for (int it : nums)
            ret += solve(nums, target, val + it);
        return dp[val] = ret;
    }

    public int combinationSum4(int[] nums, int target) {
        dp = new int[target];
        Arrays.fill(dp, -1);

        return solve(nums, target, 0);
    }
}